analyze/aps-schedule.c

```diff
@@ -0,0 +1,2989 @@
+                                     ChunkGraph* chunk_graph,
+                                     CTO_NODE* current,
+                                     CONDITION cond,
+                                     CHILD_PHASE* prev_group,
+                                     CHILD_PHASE* prev_parent,
+                                     TOTAL_ORDER_STATE* state) {
+  if (current == NULL)
+    return;
+
+  CHILD_PHASE* current_group = &current->child_phase;
+
+  if (IS_VISIT_MARKER(current)) {
+    if (current->child_phase.ch == -1) {
+      // End of total order
+      if (current->cto_next == NULL)
+        return;
+
+      // Boolean indicating whether end of phase visit marker was preceded
+      // by parent synthesized attributes <ph,-1> if any
+      bool preceded_by_parent_synthesized_current_phase =
+          prev_parent->ph > 0 && prev_parent->ch == -1;
+
+      if (aug_graph_contains_phase(aug_graph, state, cond, current_group->ph,
+                                   -1) &&
+          !preceded_by_parent_synthesized_current_phase) {
+        fatal_error(
+            "[%s] Expected parent synthesized attribute <%+d,%+d> precede the "
+            "end of phase visit marker <%+d,%+d>",
+            aug_graph_name(aug_graph), current_group->ph, -1, current_group->ph,
+            -1);
+      }
+
+      // Boolean indicating whether followed by inherited attribute of
+      // parent <-(ph+1),-1>
+      bool followed_by_parent_inherited_next_phase = false;
+      followed_by(current->cto_next, -(current->child_phase.ph + 1), -1,
+                  false /* not immediate */, false /* not visit markers only*/,
+                  &followed_by_parent_inherited_next_phase);
+
+      if (aug_graph_contains_phase(aug_graph, state, cond,
+                                   -(current_group->ph + 1), -1) &&
+          !followed_by_parent_inherited_next_phase) {
+        aps_warning(current,
+                    "[%s] Expected after end of phase visit marker "
+                    "<%+d,%+d> to be "
+                    "followed by parent inherited attribute of next "
+                    "phase <%+d,%+d>",
+                    aug_graph_name(aug_graph), current_group->ph, -1,
+                    current_group->ph + 1, -1);
+      }
+    } else {
+      // Boolean indicating whether visit marker was followed by child
+      // synthesized attribute(s) <ph,ch>
+      bool followed_by_child_synthesized = false;
+      followed_by(current->cto_next, current->child_phase.ph,
+                  current->child_phase.ch, true /* immediate */,
+                  false /* not visit markers only*/,
+                  &followed_by_child_synthesized);
+
+      // Boolean indicating whether visit marker was followed by child
+      // inherited attribute(s) <-ph,ch>
+      bool preceded_by_child_inherited = prev_group->ph == -current_group->ph &&
+                                         prev_group->ch == current_group->ch;
+
```

Comment:
Not sure how you handle the case when the visit has no attributes at all or only syntax or only int

Reply:
It's already handled I think. Local attributes don't affect `prev_group` in the recursive call and we check if there is indeed an attribute instance of `<-ph,ch>` in the instances of augmented dependency graph **and then** we throw an exception. So its completely okay for child visit to be empty OR not have inherited or synthesized attributes OR both.